firebase-firestore/src/test/java/com/google/firebase/firestore/local/IndexFreeQueryEngineTest.java

```diff
@@ -0,0 +1,281 @@
+              Iterable<DocumentKey> keys = getDocumentsInvocation.getArgument(0);
+
+              ImmutableSortedMap<DocumentKey, MaybeDocument> docs =
+                  DocumentCollections.emptyMaybeDocumentMap();
+              for (DocumentKey key : keys) {
+                docs = docs.insert(key, existingQueryResults.get(key));
+              }
+
+              return docs;
+            })
+        .when(localDocumentsView)
+        .getDocuments(any());
+
+    doAnswer(
+            getDocumentsMatchingQueryInvocation -> {
+              Query query = getDocumentsMatchingQueryInvocation.getArgument(0);
+              SnapshotVersion snapshotVersion = getDocumentsMatchingQueryInvocation.getArgument(1);
+
+              assertEquals(
+                  "Observed query execution mode did not match expectation",
+                  expectIndexFreeExecution,
+                  !SnapshotVersion.NONE.equals(snapshotVersion));
+
+              ImmutableSortedMap<DocumentKey, MaybeDocument> matchingDocs =
+                  DocumentCollections.emptyMaybeDocumentMap();
+
+              for (Document doc : updatedQueryResults.values()) {
+                if (query.matches(doc)) {
+                  matchingDocs = matchingDocs.insert(doc.getKey(), doc);
+                }
+              }
+
+              return matchingDocs;
+            })
+        .when(localDocumentsView)
+        .getDocumentsMatchingQuery(any(), any());
+  }
+
+  /** Add a document to local cache and the remote key mapping. */
+  private void addExistingResult(Document doc) {
+    existingQueryResults.put(doc.getKey(), doc);
+  }
+
+  /** Add a document to local cache but not the remote key mapping. */
+  private void addUpdatedResult(Document doc) {
+    updatedQueryResults.put(doc.getKey(), doc);
+  }
+
+  private ImmutableSortedMap<DocumentKey, Document> runQuery(
+      Query query, QueryData queryData, boolean expectIndexFree) {
+    expectIndexFreeExecution = expectIndexFree;
+    ImmutableSortedSet<DocumentKey> remoteKeys =
+        new ImmutableSortedSet<>(
+            new ArrayList<>(existingQueryResults.keySet()), DocumentKey.comparator());
+    return queryEngine.getDocumentsMatchingQuery(query, queryData, remoteKeys);
+  }
+
+  @Test
+  public void usesTargetMappingForInitialView() {
+    Query query = query("coll").filter(filter("matches", "==", true));
+    QueryData queryData = queryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    addExistingResult(MATCHING_DOC_A);
+    addExistingResult(MATCHING_DOC_B);
+
+    ImmutableSortedMap<DocumentKey, Document> docs =
+        runQuery(query, queryData, /* expectIndexFree= */ true);
+    assertEquals(asList(MATCHING_DOC_A, MATCHING_DOC_B), values(docs));
+  }
+
+  @Test
+  public void filtersNonMatchingInitialResults() {
+    Query query = query("coll").filter(filter("matches", "==", true));
+    QueryData queryData = queryData(query, /* hasLimboFreeSnapshot= */ true);
+
+    addExistingResult(NON_MATCHING_DOC_A);
```

Comment:
How would this happen? Wouldn't an existing result from the server have to match?

It seems like maybe I don't understand what an existing result is. See comment above.